src/logger.ts

```diff
@@ -1,8 +1,56 @@
 /**
- * A method that returns the default logger for this application.
- * Uses a function instead of a constant to allow mocking.
- * @returns The default logger for this application
+ * The default logger for this application.
+ * Can be set to any object that has `debug`, `info`, `warn`, and `error` methods,
+ * as long as it can be cast to a `Console`.
+ * @public
  */
-export function getLogger(): Console {
-	return console;
+const logger: Console = console;
+export default logger;
```

Comment:
Considering the entire module is for the logger, and all sub-exports are sub-methods of the logger, it makes more sense to me to have the default export (IE, importing the entire module) simply be the logger.

It's similar to how discord.js works. You can import the entire module as the Discord object (`import Discord from 'discord.js'`), or import sub-classes of Discord (`import { Client } from 'discord.js'`).

It wouldn't make sense to have the `logger` object be on the same export level as its methods (`export { logger, debug, info, warn, error }`) when the methods belong to the logger (`logger.debug, logger.info, ...`). Can you see importing the entire module as the logger object makes more sense, when the logger object IS the module?

Reply:
We aren't exporting the entire module here though, we're exporting an object named `default`, which JS lends some syntactic sugar to let importers name the import whatever they want, transparently. TypeScript does some translation work to facilitate that import in the compiled output, (which you can read in `dist/main.js`), which work TS omits as unnecessary with named exports.

Discord.js doesn't use `export default`, at least not at the top level. They use named exports, in the CommonJS `exports.name` syntax, the same syntax to which TS translates our named imports. TS translates our import into a CommonJS `require` in the same way that it would translate an `import * as package from 'package'`. You can observe this by adding the following to `main.ts`, running `npm run build:only`, and observing `dist/main.js`:

```ts
import * as _Discord from 'discord.js';
import Discord from 'discord.js';
```

To really "import the entire module," importers should, IMO, use `import * as logger from './logger'` syntax. I'd prefer even that to default export, whose implicit renaming creates hidden room for inconsistent spellings and partial refactors, where the `as` operator or the `import { name }` syntax make clear that a rename is going on.

There's nothing I can do to prevent us using packages that rely on `export default`, that's their decision to make and our code can interop with that just fine. However, for the reasons given in [this article](https://basarat.gitbook.io/typescript/main-1/defaultisbad) (esp. good-practice CommonJS interop, consistent autocomplete, and auto-import), I plan to update our ESLint config to enforce the use of named exports in our code.

I agree that these parts shouldn't be on the same export level. Perhaps there can be some compromise? I'll approve the code as-is, but if it's merged, I intend to try a few things later to factor out that default export.

Reply:
(I realize now that my comment has "If I may interject for a moment" vibes, but about the merits of named vs. `default` exports, lol)

Reply:
TBH, I mostly don't want us falling into the habit of using `export default` everywhere. I'd be plenty happy for now if you followed this SO answer to disallow `export default` by default, then add an `eslint-disable-next-line` marker in `logger.ts` to whitelist its use here.

Reply:
Ah, my bad, it appears I didn't understand TS exports as well as I thought.

Is there an alternative solution to organize the exports of `logger.ts` better (without using unbounded methods)?

Reply:
There are a couple of ways. For one thing, unbound methods are not an issue in a test environment if the method isn't really a function, but a `jest.Mock` instance, and if TypeScript properly knows that. You _could_ export a named `logger` with the properties of `Console`, or at least a basic interface. The version in `__mocks__` would do the same, except its methods would all be `jest.fn()`. Then in the test, calling `jest.mock('./logger')` and then importing the logger would return that mocked instance. (You know this.) To reference the methods, you'd do something like this:
```ts
// *.test.ts
jest.mock('./logger');
import { logger } from './logger';
const mockLoggerError = logger.error as jest.Mock;

// ...
expect(mockLoggerError).toHaveBeenCalledWith(...);
```
This doesn't \*really\* reference an unbound method, since TypeScript knows `jest.Mock` will never use `this`, and this holds true at runtime since you've mocked the module. The extra typecast just tells TypeScript what's going to happen at runtime in this test. This way, you never need to export the method functions in either `logger.ts` or its mock, just that named `logger` object.

There are other ways to skin this cat as well:
- Export `logger` on the same level as its methods, import the methods directly in tests
- Remove the default export and use `import * as logger` to use the logger. This would glob together the methods defined in `logger.ts` (which are exported by name at the moment) into a new `logger` interface, a subset of `Console`, and one that we always know is mocked. We don't need to worry about adding mocks for other `Console` APIs, because there would be no `Console` APIs exported! (This also leaves room for naming the `logger` interface something else, but we can fix that in other ways. At least finding such cases would be as easy as searching for "import * as` in the codebase.)
- Probably other ways that I'm too tired to think up right now

Reply:
Alas, the unbound methods issue remains
![image](https://user-images.githubusercontent.com/28303477/194430894-b9184c34-49c2-404c-abc2-3be586b6964d.png)

Reply:
And it'd be very messy to put `// eslint-disable-next-line @typescript-eslint/unbound-method` in every test script that uses the logger...

Reply:
Your last two suggestions are actually mutually inclusive - we can do both at the same time. That'd also allow us to define our own logger interface with any number of endpoints (if we so wished) and allow `logger.ts` to take care of redirecting the output.

Reply:
I must have been mistaken about that unbound-method issue. I wonder if we can configure `eslint-plugin-jest` to override the `@typescript-eslint` error when it makes sense, something like [this](https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/unbound-method.md), perhaps. Might even consider disabling `@typescript-eslint/unbound-method` entirely for test files.

You're right tho, if we export our own functions and import with `import * as`, then there'd be no unbound method problem.

Reply:
I tried that jest eslint rule, and it didn't work.
`'@typescript-eslint/unbound-method': 'off'` disabled the error, but
`'jest/unbound-method': 'error'` just threw the error instead.
Apparently it couldn't accurately identify the casting as safe.

Reply:
And yeah, I disabled `@typescript-eslint/unbound-method` for test files at first, and that definitely worked.

Reply:
Fine by me. I don't anticipate we'll be doing much with actual unbound methods, just these cases with test files where the method is mocked

Reply:
I think I'll try the `import *` method. I like the idea of separating our implementation of `logger` from the interface of `Console`, so in the future, we could change the logger to perform any behavior we wish, independent of the Console endpoints.

(Plus that'll reduce the number of lines needed for exporting and importing, and I wanna improve my +/- ratio)